core/app/models/spree/address.rb

```diff
@@ -157,11 +157,17 @@ def country_iso=(iso)
 
     private
       def state_validate
-        # Skip state validation without country (also required)
-        # or when disabled by preference
-        return if country.blank? || !Spree::Config[:address_requires_state]
-        return unless country.states_required
+        # Skip state validation without country
+        return if country.blank?
 
+        # If states are not required we still check if state belongs to country
+        # If it doesn't we just remove it without adding state errors
+        unless country.states_required || readonly?
```

Comment:
I've added the readonly check thinking it was just some special case, not knowing about the immutability part since our store is still on 1.0. However, if we remove it all specs pass except this one. 
```
  1) Spree::CheckoutController#update fails to transition to complete from confirm when the country is not a shippable country due to no available shipping rates for any of the shipments
     Failure/Error: order.update_attributes(state: 'confirm')
     ActiveRecord::ReadOnlyRecord:
       Spree::Address is marked as readonly
     # ./spec/controllers/spree/checkout_controller_spec.rb:358:in `block (5 levels) in <top (required)>'
``` which doesn't seem related at all.

Putting in the merge function doesn't seem explicit it at all, and the validation still runs when creating the new address.